From 686f2f005bc196d36c2c90395d69720a6f4f31e6 Mon Sep 17 00:00:00 2001
From: Valentin Raevsky <valentin@compulab.co.il>
Date: Mon, 24 Oct 2022 17:23:44 +0300
Subject: [PATCH] imx8mp: config: Introduced D1D8/D2D4 configs
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Introduced D1D8 and D2D4 dram configurations.
The SoC SPL code can’t contain all available CompuLab dram
configurations due to the limited ocram size 576K. That is why the dram
timings code was broken down into two subsets that support differed dram
capacities: D1D8 – 1G and 8G; D2D4 – 2G and 4G.

Default configuration is D1D8.

1) Enable D1D8:
make O=build ucm-imx8m-plus_defconfig
cat << eof > build/d1d8.cfg
# CONFIG_DRAM_D2D4 is not set
eof
KCONFIG_CONFIG=build/.config  ./scripts/kconfig/merge_config.sh -m
build/.config build/d1d8.cfg
2) Enable D2D4:
make O=build ucm-imx8m-plus_defconfig
cat << eof > build/d2d4.cfg
CONFIG_DRAM_D2D4=y
eof
KCONFIG_CONFIG=build/.config  ./scripts/kconfig/merge_config.sh -m
build/.config build/d2d4.cfg

Signed-off-by: Valentin Raevsky <valentin@compulab.co.il>
---
 board/compulab/plat/imx8mp/Kconfig       |  7 +++---
 board/compulab/plat/imx8mp/board/board.c | 14 +++++------
 board/compulab/plat/imx8mp/ddr/Makefile  | 11 ++++-----
 board/compulab/plat/imx8mp/ddr/ddr.h     | 30 +++++++++++++-----------
 4 files changed, 32 insertions(+), 30 deletions(-)

diff --git a/board/compulab/plat/imx8mp/Kconfig b/board/compulab/plat/imx8mp/Kconfig
index cd05159112..0ccf5ac9b9 100644
--- a/board/compulab/plat/imx8mp/Kconfig
+++ b/board/compulab/plat/imx8mp/Kconfig
@@ -1,10 +1,11 @@
 if TARGET_SOM_IMX8M_PLUS || TARGET_UCM_IMX8M_PLUS || TARGET_IOT_GATE_IMX8PLUS
 
-config DRAM_SEC_SUBSET
-	bool "Enable Second DRAM subset"
+config DRAM_D2D4
+	bool "Enable D2,D4 or D1,D8 dram support"
 	default n
 	help
-	  This option enables Samsung 1G and 4G dram support
+	  If 'Y' then support for D2 and D4 configuration is enabled
+	  else support for D1 and D8 configuration is enabled
 
 config ATF_LOAD_ADDR
 	hex "bl31.bin load address"
diff --git a/board/compulab/plat/imx8mp/board/board.c b/board/compulab/plat/imx8mp/board/board.c
index 6f5b94a1d2..e3218c0666 100644
--- a/board/compulab/plat/imx8mp/board/board.c
+++ b/board/compulab/plat/imx8mp/board/board.c
@@ -515,10 +515,10 @@ int board_late_init(void)
 	env_set("board_rev", "iMX8MP");
 #endif
 
-#ifdef CONFIG_DRAM_SEC_SUBSET
-	env_set("dram_subset", "2");
+#ifdef CONFIG_DRAM_D2D4
+	env_set("dram_subset", "d2d4");
 #else
-	env_set("dram_subset", "1");
+	env_set("dram_subset", "d1d8");
 #endif
 	return 0;
 }
@@ -730,15 +730,15 @@ int do_ddr(struct cmd_tbl *cmdtp, int flag, int argc, char *const argv[])
    return 0;
 }
 
-#ifdef CONFIG_DRAM_SEC_SUBSET
-#define VERSION "2"
+#ifdef CONFIG_DRAM_D2D4
+#define SUPPORTED_CONF "D2,D4"
 #else
-#define VERSION "1"
+#define SUPPORTED_CONF "D1,D8"
 #endif
 
 U_BOOT_CMD(
 	ddr,	2,	1,	do_ddr,
-	"rdmr/read/clear\nversion : [ "VERSION" ]" ,
+	"rdmr/read/clear\nSupported configurations : [ "SUPPORTED_CONF" ]" ,
 	ddr_help_text
 );
 #endif
diff --git a/board/compulab/plat/imx8mp/ddr/Makefile b/board/compulab/plat/imx8mp/ddr/Makefile
index e27edb7c69..78574f89a7 100644
--- a/board/compulab/plat/imx8mp/ddr/Makefile
+++ b/board/compulab/plat/imx8mp/ddr/Makefile
@@ -1,12 +1,11 @@
-obj-$(CONFIG_SPL_BUILD) += lpddr4_timing_01061010_2G.o
-
-ifeq ($(CONFIG_DRAM_SEC_SUBSET),y)
+ifeq ($(CONFIG_DRAM_D2D4),y)
+obj-$(CONFIG_SPL_BUILD) += lpddr4_timing_ff000010.o
 obj-$(CONFIG_SPL_BUILD) += lpddr4_timing_01061010_4G.o
-obj-$(CONFIG_SPL_BUILD) += lpddr4_timing_01061010_1G.o
-obj-$(CONFIG_SPL_BUILD) += lpddr4_timing_01061010_1G_4000.o
+obj-$(CONFIG_SPL_BUILD) += lpddr4_timing_01061010_2G.o
 else
-obj-$(CONFIG_SPL_BUILD) += lpddr4_timing_ff000010.o
 obj-$(CONFIG_SPL_BUILD) += lpddr4_timing_ff060018.o
+obj-$(CONFIG_SPL_BUILD) += lpddr4_timing_01061010_1G.o
+obj-$(CONFIG_SPL_BUILD) += lpddr4_timing_01061010_1G_4000.o
 endif
 obj-$(CONFIG_SPL_BUILD) += ddr.o
 obj-y += ddr_common.o
diff --git a/board/compulab/plat/imx8mp/ddr/ddr.h b/board/compulab/plat/imx8mp/ddr/ddr.h
index ff09db8ed6..869bbb4689 100644
--- a/board/compulab/plat/imx8mp/ddr/ddr.h
+++ b/board/compulab/plat/imx8mp/ddr/ddr.h
@@ -8,13 +8,13 @@
 #define __DDR_H__
 
 #ifdef CONFIG_SPL_BUILD
-#ifdef CONFIG_DRAM_SEC_SUBSET
+#ifdef CONFIG_DRAM_D2D4
+extern struct dram_timing_info ucm_dram_timing_01061010_2G;
+extern struct dram_timing_info ucm_dram_timing_ff000010;
 extern struct dram_timing_info ucm_dram_timing_01061010_4G;
+#endif
 extern struct dram_timing_info ucm_dram_timing_01061010_1G;
 extern struct dram_timing_info ucm_dram_timing_01061010_1G_4000;
-#endif
-extern struct dram_timing_info ucm_dram_timing_01061010_2G;
-extern struct dram_timing_info ucm_dram_timing_ff000010;
 extern struct dram_timing_info ucm_dram_timing_ff060018;
 #endif
 
@@ -46,18 +46,22 @@ struct lpddr4_desc {
 };
 
 static const struct lpddr4_desc lpddr4_array[] = {
-	{ .name = "Samsung",	.id = 0xDEADBEEF, .subind = 0x2, .size = 2048, .count = 1,
+#ifdef CONFIG_DRAM_D2D4
+	{ .name = "deadbeaf",	.id = 0xdeadbeef, .subind = 0x2, .size = 2048, .count = 1,
 #ifdef CONFIG_SPL_BUILD
 		.timing = &ucm_dram_timing_01061010_2G
 #endif
 	},
-#ifdef CONFIG_DRAM_SEC_SUBSET
 	{ .name = "Samsung",	.id = 0x01061010, .subind = 0x4, .size = 4096, .count = 1,
 #ifdef CONFIG_SPL_BUILD
 		.timing = &ucm_dram_timing_01061010_4G
 #endif
 	},
+	{ .name = "Micron",	.id = 0xff000010, .subind = 0x4, .size = 4096, .count = 1,
+#ifdef CONFIG_SPL_BUILD
+		.timing = &ucm_dram_timing_ff000010
 #endif
+	},
 	{ .name = "Nanya",	.id = 0x05000010, .subind = 0x2, .size = 2048, .count = 1,
 #ifdef CONFIG_SPL_BUILD
 		.timing = &ucm_dram_timing_01061010_2G
@@ -68,22 +72,20 @@ static const struct lpddr4_desc lpddr4_array[] = {
 		.timing = &ucm_dram_timing_01061010_2G
 #endif
 	},
-#ifdef CONFIG_DRAM_SEC_SUBSET
-	{ .name = "Samsung",	.id = 0x01050008, .subind = 0x1, .size = 1024, .count = 1,
+#else
+	{ .name = "deadbeaf",	.id = 0xdeadbeaf, .subind = 0x1, .size = 1024, .count = 1,
 #ifdef CONFIG_SPL_BUILD
 		.timing = &ucm_dram_timing_01061010_1G
 #endif
 	},
-	{ .name = "Samsung",	.id = 0x01060008, .subind = 0x1, .size = 1024, .count = 1,
+	{ .name = "Samsung",	.id = 0x01050008, .subind = 0x1, .size = 1024, .count = 1,
 #ifdef CONFIG_SPL_BUILD
-		.timing = &ucm_dram_timing_01061010_1G_4000
+		.timing = &ucm_dram_timing_01061010_1G
 #endif
 	},
-#endif
-#ifndef CONFIG_DRAM_SEC_SUBSET
-	{ .name = "Micron",	.id = 0xff000010, .subind = 0x4, .size = 4096, .count = 1,
+	{ .name = "Samsung",	.id = 0x01060008, .subind = 0x1, .size = 1024, .count = 1,
 #ifdef CONFIG_SPL_BUILD
-		.timing = &ucm_dram_timing_ff000010
+		.timing = &ucm_dram_timing_01061010_1G_4000
 #endif
 	},
 	{ .name = "Micron",	.id = 0xff060018, .subind = 0x8, .size = 8192, .count = 1,
-- 
2.17.1

